### ISSUE TITLE
Media quota system

### ISSUE THREAD

**Issue by Gargron** [tootsuite/mastodon#11808]
### Pitch

Track media usage by account and limit future uploads once it's crossed.

### Motivation

Accountability for disk usage.

**Comment by glubsy** [comment]
That's actually an interesting approach to the problem.

**Comment by wizzwizz4** [comment]
Maybe quotas could be initially limited, but as the account ages / is used it gets raised to the _actual_ limit. That'd help work around the extra accounts problem slightly (at least, as applicable to people using it as a file-hosting service).

This might not be a good plan for a base feature, though.

**Comment by glubsy** [comment]
Brilliant thinking @wizzwizz4 ! Definitely a good solution.

**Comment by AstroProfundis** [comment]
It would be a good feature for instance moderators, but I don't think the main problem of disk usage is about user uploaded fancy pictures of the local instance but cached media files from the whole Fediverse.

There're issues like #9567 and #10049 discusting about similar topics.

**Comment by ldexterldesign** [comment]
Fixed quota vs. relative quota (props @wizzwizz4) options seems like a good first step

For example:

* Fixed quota x could be 100MB
* Relative quota x could be equal to days since account creation (i.e. account 100 days old then quota is 100MB)

My real world metaphor is how my ISP handles bandwidth throttling (i.e. if my quota is 20GB per day then if I download >20GB per day or >140GB per week or >608GB per month etc. then I get limited)

If users delete toots with attachments then this data gets reallocated to their quota or, for convenience, there could be a feature to auto-delete toots >x days old

Lovely initial concept framing @gargron!

Hope this is useful

Regards

**Comment by wizzwizz4** [comment]
@ldexterldesign I don't like the "autodelete toots" idea all that much, but I am generally wary of accidental loss of important data. Perhaps it could be a per-toot thing, turned off by default and adjustable after posting.

The monotonously-increasing quota seems like a good idea, though I fear it could just be abused by leaving an account open for a few months before it'll be needed for file storage.

**Comment by ldexterldesign** [comment]
@wizzwizz4

> [...] turned off by default

Yes, totally. If a user's new post will exceed quota then info' on how to fix this which includes a mention of auto-delete feature makes sense to me.

> The monotonously-increasing quota seems like a good idea, though I fear it could just be abused by leaving an account open for a few months before it'll be needed for file storage.

Without subjectively evaluating the agency (e.g. follow/follower ratio, boosts) of accounts I can't think of another objective metric than time (e.g. days registered) to aid an algorithm of this type. I would argue that the act of having a quota should prevent quota abuse. File storers gonna store.

![foo](https://user-images.githubusercontent.com/1715809/67253132-76534e00-f476-11e9-887d-4e38343a6d91.gif)

Regards

**Comment by koyuawsmbrtn** [comment]
I would like to add that we could display media usage in "blocks" like on the Nintendo 3DS. Let's say we would have a max quota of 5000 blocks and if you want more blocks you just go and pay more or ask the admin to stack up the blocks for everyone.

**Comment by Gargron** [comment]
@koyuawsmbrtn What do 5000 blocks mean? Why blocks and not MB?

**Comment by koyuawsmbrtn** [comment]
> @koyuawsmbrtn What do 5000 blocks mean? Why blocks and not MB?

5000 blocks equal 5000x8MB=40000MB=around 40GB and I thought using blocks was more accessible to normies.

**Comment by ldexterldesign** [comment]
Hi @koyuawsmbrtn,

Hmm, if I understand correctly then it's gonna get confusing if a block on one instance can mean something different to a block on another instance. It's 2020, I think there's more chance of someone understanding what a megabyte is over an arbitrary block.

Also, it would be a personal turn off for me to see instances monetising disk space ala private torrent trackers

Regards

**Comment by Eriner** [comment]
Remote media (and preview images, and all other remote files) would ideally have both a global and per-server configurable storage quota for both duration and size. For a concrete example, consider the desired implementation below (read top-to-bottom like ACLs or firewall rules):

```
* - [remote_media, preview_images]: limit 90d
* - [remote_media, preview_images]: limit 25G
imagespam.example - [remote_media, preview_images]: limit 10G
linkspam.example - [preview_images]: limit 10G
twitter.com - [*]: limit 14d
twitter.com - [*]: limit 100G
```

This would enable automated storage cleanup and constraining disk usage to a (more) expected range. Being able to define and select multiple storage backends (for "hotter" and "colder" storage) would be nice, but I'll save that for another time.

Many cron-based administrative tasks (`tootctl media remove --days=x`) would become administrator/operator configurable settings.

These controls would be very useful, particularly in the event that high-volume websites like Imgur, Twitter, YouTube, Reddit, etc. suddenly enable federation.
